### ISSUE TITLE
Emscripten compilation issue

### ISSUE THREAD

**Issue by blackccpie** [tiny-dnn/tiny-dnn#854]
Hi,

I'm having troubles compiling my tiny-dnn based applications with emscripten since the xtensor integration.
The error is easily reproducible with the following code sample:

```C++
#include "tiny_dnn/tiny_dnn.h"

int main( int argc, char **argv )
{
    using namespace tiny_dnn;
    network<sequential> nn;
    return 0;
}
```

I then try to compile this with the following *CMakeLists.txt* file:
```cmake
cmake_minimum_required (VERSION 3.0)

project(tinytest)

set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

include_directories(../tiny-dnn)

add_executable(tinytest main.cpp)

target_link_libraries(tinytest
    pthread
)
```
and I launch the build using these command lines:

```sh
$ mkdir build
$ cd build
$ cmake -DCMAKE_TOOLCHAIN_FILE=$EMSCRIPTEN/cmake/Modules/Platform/Emscripten.cmake -DUSE_EMSCRIPTEN=ON ..
$ make
```

The *EMSCRIPTEN* environment variable is defined after having sourced the following file in the emscripten emsdk directory:

```sh
$ . ./emsdk_env.sh
```

The error I'm getting is related to ambiguous call to *is_trivially_default_constructible*:
```
tiny-dnn/tiny_dnn/xtensor/xutils.hpp:960:64: error: unknown template name 'is_trivially_default_constructible'
        struct xtrivial_default_construct_impl<true, T> : std::is_trivially_default_constructible<T> {};
                                                               ^
tiny-dnn/tiny_dnn/xtensor/xutils.hpp:967:103: error: reference to 'is_trivially_default_constructible' is ambiguous
    using xtrivially_default_constructible = detail::xtrivial_default_construct_impl<is_complete<std::is_trivially_default_constructible<double>>::value, T>;
                                                                                                 ~~~~~^
emsdk-portable/emscripten/1.37.9/system/include/libcxx/type_traits:3418:51: note: candidate found by name lookup is 'std::__2::is_trivially_default_constructible'
template <class _Tp> struct _LIBCPP_TYPE_VIS_ONLY is_trivially_default_constructible
                                                  ^
tiny-dnn/tiny_dnn/xtensor/xutils.hpp:949:43: note: candidate found by name lookup is 'std::is_trivially_default_constructible'
namespace std { template <class T> struct is_trivially_default_constructible; }
                                          ^
tiny-dnn/tiny_dnn/xtensor/xutils.hpp:967:103: error: unknown type name 'is_trivially_default_constructible'
    using xtrivially_default_constructible = detail::xtrivial_default_construct_impl<is_complete<std::is_trivially_default_constructible<double>>::value, T>;
                                                                                                      ^
tiny-dnn/tiny_dnn/xtensor/xutils.hpp:967:157: error: expected a type
    using xtrivially_default_constructible = detail::xtrivial_default_construct_impl<is_complete<std::is_trivially_default_constructible<double>>::value, T>;
```

When compiling with standard gcc I get no errors.

Could someone help me on this?

thx

PS: I'm running in an Ubuntu 16.04 Linux, and using emscripten sdk version 1.37.9

**Comment by blackccpie** [comment]
One quick and dirty workaround is to comment this forward declaration in *xtensor/xutils.hpp* l.949:
```C++
namespace std { template <class T> struct is_trivially_default_constructible; }
```
I then ran into another compiling error, where I had to add this std::forward call in *util/serialization_function.h* l.46:
```C++
cereal::NameValuePair<serial_size_t> arg2(arg.name, std::forward<cereal::NameValuePair<size_t>>(arg).value);
```

**Comment by Randl** [comment]
cc @SylvainCorlay

**Comment by SylvainCorlay** [comment]
Now regarding how you can work around this, maybe emscripten tries to instantiate all forward-declared templates even when they are actually not used.

Maybe we can put these forward declarations behind a flag that could be enabled in the cmake call.

Ping @maartenbreddels who is also interested in using xtensor in the browser.

**Comment by blackccpie** [comment]
@SylvainCorlay thanks for the detailed explanations.

Emscripten always defines the `__EMSCRIPTEN__` preprocessor macro, so perhaps we could modify *xutils.hpp* l.943 like this:

```C++
#if !(defined(__APPLE__)) && !(defined(__EMSCRIPTEN__))
```
It compiles fine on my machine, in fact it brings the Emscripten case similar to the Apple+Clang case.

**Comment by SylvainCorlay** [comment]
Sounds like a good fix. PR welcome in xtensor!

**Comment by blackccpie** [comment]
@SylvainCorlay : No problem, I'll handle the PR.
@Randl : are you ok with the *serialization_function.h* correction? I can PR too

**Comment by SylvainCorlay** [comment]
PS: xtensor exists out of tiny-dnn: https://github.com/QuantStack/xtensor/

**Comment by blackccpie** [comment]
@SylvainCorlay : I'll PR on the QuantStack repo ;-)

**Comment by Randl** [comment]
yeah, I think it's fine

**Comment by procoru** [comment]
Today I was able to build a [simple c++ application](https://github.com/procoru/tiny-dnn-to-wasm/blob/master/main.cpp) using tiny-dnn in a [wasm file](https://github.com/procoru/tiny-dnn-to-wasm/blob/master/main.wasm) for use in a browser from js.

All laid out raw [here](https://github.com/procoru/tiny-dnn-to-wasm).

**Comment by blackccpie** [comment]
@procoru : thx for sharing your experience. It's strange you didn't run into the same compiling issues as I did. Maybe it's related to your linux os version.

With older tiny-dnn code versions I managed to run some successfull experiments. If you interested in the topic you can also browse my repo [here](https://github.com/blackccpie/tiny-brain) or have a look at the demo [here](http://blackccpie.free.fr/ocr/) with an up-to-date firefox or chrome.

**Comment by bhack** [comment]
@procoru Nice! /cc @edgarriba

**Comment by bhack** [comment]
@procoru Do you want to propose a PR to integrate the process upstream?

**Comment by blackccpie** [comment]
This issue is halfway corrected with the help of #855 being now merged. The other missing half would be to merge the aforementioned xtensor PR [#388](https://github.com/QuantStack/xtensor/pull/388) into tiny-dnn head, or to merge the latest xtensor release. Since I don't known how the project owners deal with xtensor updating, I'm asking for the correct way to achieve this? thx
